### PR TITLE
perf(cache): batch foreground blob lookups

### PR DESCRIPTION
**Pull request by jdx**
## Summary
- add a batched FindBlobs request to the task-scoped cache agent
- use blob-pack downloads for foreground rustc restore blob batches when the remote supports them
- keep individual blob GET fallback through the existing remote client path

## Tests
- cargo check --locked -p mise-cache-core --all-features
- cargo test --locked -p mise-cache-core blob_pack
- cargo test --locked --bin mise cache::rustc
- cargo fmt --all -- --check

*AI-assisted — Tool: Codex; model: unavailable; version: unavailable.*

---

> [!NOTE]
> **Medium Risk**
> Touches the cache agent wire protocol and shared remote blob download logic used by rustc restore and prefetch; changes are additive (single-blob API unchanged) and covered by blob-pack tests, but incorrect batching or semaphore behavior could affect restore latency or concurrency.
> 
> **Overview**
> **Batched blob resolution** for foreground cache restores: the agent protocol gains `FindBlobs` / `Blobs`, and the rustc shim issues one agent request per multi-digest lookup instead of one `FindBlob` per digest, preserving input order (including duplicates) in the response.
> 
> **Remote fetch path** is generalized: `prefetch_remote_blobs` becomes `fetch_remote_blobs` with an optional prefetch semaphore. Prefetch callers pass `Some(&prefetch_transfers)`; foreground `find_blobs` passes `None` so restores can use blob packs without competing for prefetch slots. Blob-pack HTTP calls now hold both prefetch (when set) and `remote_transfers` permits, with transfer timing recorded outside the permit scope on failure.
> 
> **Tests** add coverage that `FindBlobs` on the read path uses a single blob-pack request and zero per-blob GETs when the remote supports packs.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 6384579597a6f222fd2f8d86fa642692ab36b9a8. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **Performance Improvements**
  * Improved handling of multiple cached files by processing blob lookups in batches.
  * Reduced redundant requests when retrieving compiler cache data.
  * Preserved request order while efficiently reusing locally verified content.

* **Reliability**
  * Improved validation and error handling for incomplete, missing, or unexpected cache responses.
  * Added coverage for foreground cache downloads.

### CODE REVIEW

File: crates/mise-cache-core/src/agent.rs [hunk 1]
```diff
@@ -1392,6 +1406,35 @@ impl CacheAgent {
...
+        {
+            paths.extend(self.fetch_remote_blobs(remote, missing, None).await);
+        }
+
+        Ok(AgentResponse::Blobs {
+            paths: digests
+                .into_iter()
+                .map(|digest| paths.remove(&digest))
+                .collect(),
+        })
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mise-cache-core/src/agent.rs:1457]: Fixed in b9fc759: batched responses now clone paths from the digest map instead of removing them, so duplicate digests resolve repeatedly. I also updated the foreground blob-pack test to request a duplicate digest.

*AI-assisted — Tool: Codex; model: unavailable; version: unavailable.*
